Add support for parsing N-Quads

nikokaoja opened the pull request.

Tpt commented on tests/data/test.nq:
```diff
@@ -0,0 +1,6 @@
+<http://example.org/stuff/1.0/document> <http://xmlns.com/foaf/0.1/creator> <http://example.org/stuff/1.0/creator> <http://example.com/graph1> .
+<http://example.org/stuff/1.0/creator> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://xmlns.com/foaf/0.1/Person> <http://example.com/graph1> .
+<http://example.org/stuff/1.0/creator> <http://xmlns.com/foaf/0.1/name> "John Doe" <http://example.com/graph1> .
+<http://example.org/stuff/1.0/creator> <http://xmlns.com/foaf/0.1/mbox> <mailto:johndoe@example.org> <http://example.com/graph1> .
```
Would be great to have a few quads in the default graph to check that they are properly loaded into the default graph

Tpt commented on tests/test_parser.py:
```diff
@@ -127,6 +133,18 @@ def test_parsing_ox_rdfxml_fallback(self):
+        graph = rdflib.Dataset(store="Oxigraph")
+        graph.parse(_TEST_DIR / "data/test.nq", format="ox-nquads", transactional=False)
+        self.assertEqual(len(graph), 6)
+        self.assertEqual(len(graph.query(_NAMEDGRAPH_QUERY)), 1)
```
Instead, it would be great to check the number of triple into the default graph and the named graphs. rdflib behaviors related to named graphs are weird and there might be bad surprises.

Tpt reviewed (commented):
Thank you! Might be great to have some test coverage around the default graph (default graph behaviors of rdflib are weird)

Tpt commented on tests/data/test.nq:
```diff
@@ -0,0 +1,6 @@
+<http://example.org/stuff/1.0/document> <http://xmlns.com/foaf/0.1/title> "Example Document" <http://example.com/graph2> .
+<http://example.org/stuff/1.0/creator> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://xmlns.com/foaf/0.1/Person> <http://example.com/graph3> .
+<http://example.org/stuff/1.0/creator> <http://xmlns.com/foaf/0.1/name> "John Doe" <urn:x-rdflib:default> .
+<http://example.org/stuff/1.0/creator> <http://xmlns.com/foaf/0.1/mbox> <mailto:johndoe@example.org> <urn:x-rdflib:default> .
```
It would be great to test with the actual NQuads default graph like  `<http://example.org/stuff/1.0/creator> <http://xmlns.com/foaf/0.1/mbox> <mailto:johndoe@example.org> .`.

I believe proper support for the default graph would require moving the quads in a named graph (likely `urn:x-rdflib:default`). Sadly pyoxigraph 0.3.x does not support that but pyoxigraph 0.4 alphas does.

Tpt commented on oxrdflib/parser.py:
```diff
@@ -47,6 +47,7 @@ def parse(
 
         else:
             base_iri = sink.absolutize(source.getPublicId() or source.getSystemId() or "")
+            graph_identifier = to_ox(sink.identifier) if format not in ("ox-nq", "ox-nquads") else None
```
We should map the default graph for dataset formats too.

```suggestion
            graph_identifier = to_ox(sink.identifier)
```

Tpt commented:
@nikokaoja I have merged pyoxigraph 0.4 support in the main branch. Do you want to rebase your MR? If not, may I build on top of it?

nikokaoja commented:
@Tpt great ! Will do update of this PR by end of this week.

Tpt commented:
@nikokaoja Amazing! I already implemented serializers in #51 and found a few ways to simplify the code a bit.

nikokaoja commented:
![Screenshot 2024-10-26 at 09 55 44](https://github.com/user-attachments/assets/97ef307f-c57c-4508-85b7-ede67ac83efa)

@Tpt still same issue when rebasing main, `default_context` is empy

Tpt commented:
Thank you for pushing this! I wrote an untested draft that  should properly insert in the correct graph and supports other store: #53

nikokaoja commented:
> Thank you for pushing this! I wrote an untested draft that should properly insert in the correct graph and supports other store: #53 Feel free to reuse the code. I also found a nice way to automate tests for serializers, it might be handy for the parsers too

Still the same issue, `.default_context` is empty

Tpt commented:
> @Tpt still the same issue, .default_context is empty. I guess it is something with config on pyoxigraph side?

This is weird, I have updated #53 with some tests that are passing. I also fixed the loading into not-Oxigraph stores that seems now to work.

nikokaoja commented:
> > @Tpt still the same issue, .default_context is empty. I guess it is something with config on pyoxigraph side?
> 
> This is weird, I have updated #53 with some tests that are passing. I also fixed the loading into not-Oxigraph stores that seems now to work.
> 
> Edit: tests are passing with Python 3.12 but not with Python 3.8. It sounds like something sensitive to the Python version.

I am running it with Python 3.11 and have the above reported behaviour. So it seems it works only with 3.12

Tpt commented:
I understood the error, it's because of changes in rdflib 7.1 that fixes the support of default graph. I am afraid it will be hard to make things work with older versions. I have gated the tests in my PR on v7.1 and also fixed a small bug in OxigraphStore in #54 that was making iteration on the dataset fail. If it's fine with you, I can maybe just merge my MR. WDYT?

krista-t commented:
> I understood the error, it's because of changes in rdflib 7.1 that fixes the support of default graph. I am afraid it will be hard to make things work with older versions. I have gated the tests in my PR on v7.1 and also fixed a small bug in OxigraphStore in #54 that was making iteration on the dataset fail. If it's fine with you, I can maybe just merge my MR. WDYT?

sure, then del this PR, I have a small comment on your PR

nikokaoja commented:
> I understood the error, it's because of changes in rdflib 7.1 that fixes the support of default graph. I am afraid it will be hard to make things work with older versions. I have gated the tests in my PR on v7.1 and also fixed a small bug in OxigraphStore in #54 that was making iteration on the dataset fail. If it's fine with you, I can maybe just merge my MR. WDYT?

Sure, np , remove my PR then.

nikokaoja commented:
This PR has been abandoned
